### PR TITLE
Dat 2875 + refactor

### PR DESCRIPTION
**Pull request by idradm**

### CODE REVIEW

File: extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+
+class ApiQueryPortableInfobox extends ApiQueryGeneratorBase {
+
+	public function __construct( $query, $moduleName ) {
+		parent::__construct( $query, $moduleName, 'ib' );
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php]: Just for clarity: why 'ib' and not something else, i.e. 'pib'?

File: extensions/wikia/PortableInfobox/controllers/PortableInfoboxParserTagController.class.php [hunk 2]
```diff
@@ -46,23 +54,28 @@ public static function replaceInfoboxMarkers( &$parser, &$text ) {
...
+	 * @param array $params
+	 *
 	 * @return string
 	 * @throws UnimplementedNodeException when node used in markup does not exists
 	 * @throws XmlMarkupParseErrorException xml not well formatted
 	 */
 	public function render( $markup, Parser $parser, PPFrame $frame, $params = null ) {
-		$infoboxParser = new Wikia\PortableInfobox\Parser\XmlParser( $this->getFrameParams( $frame ) );
-		$infoboxParser->setExternalParser( new Wikia\PortableInfobox\Parser\MediaWikiParserService( $parser, $frame ) );
+		$infoboxNode = Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML(
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/controllers/PortableInfoboxParserTagController.class.php]: Could we use just Nodes\NodeFactory here?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 3]
```diff
@@ -55,10 +42,17 @@ public function getExternalParser() {
...
 	/**
-	 * @param mixed $externalParser
+	 * @param ExternalParser|null $externalParser
+	 *
+	 * @return $this
 	 */
-	public function setExternalParser( ExternalParser $externalParser ) {
-		$this->externalParser = $externalParser;
+	public function setExternalParser( $externalParser ) {
+		// we can pass anything, and ignore it if not ExternalParser instance
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Why was that needed?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 4]
```diff
@@ -71,8 +65,25 @@ public function getType() {
 		return strtolower( $this->xmlNode->getName() );
 	}
 
+	public function isType( $type ) {
+		return strcmp( $this->getType(), strtolower( $type ) ) == 0;
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: A nitpick as we don't use diacritics in types at the moment: wouldn't it be safer to use mb_strtolower instead of strtolower?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 5]
```diff
@@ -83,10 +94,51 @@ public function getData() {
 	 *
 	 * @return bool
 	 */
-	public function isEmpty( $data ) {
-		$value = $data[ 'value' ];
+	public function isEmpty() {
+		$data = $this->getData()[ 'value' ];
+
+		return !( isset( $data ) ) || ( empty( $data ) && $data != '0' );
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Isn't this equivalent to purely:
 return ( empty( $data ) && $data != '0' );

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php [hunk 6]
```diff
@@ -83,10 +94,51 @@ public function getData() {
...
+			return [
+				'type' => $item->getType(),
+				'data' => $item->getData(),
+				'isEmpty' => $item->isEmpty(),
+				'source' => $item->getSource()
+			];
+		}, $this->getChildNodes() );
+	}
+
+	protected function getRenderDataForChildren() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: This reads a bit hard because of formatting and using anonymous functions. 
Can we reformat so each param of array_map starts with the same level of indentation?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFactory.class.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	/**
+	 * @param \SimpleXMLElement $xmlNode
+	 * @param array $data
+	 *
+	 * @return Node|NodeUnimplemented
+	 */
+	protected static function getInstance( \SimpleXMLElement $xmlNode, array $data ) {
+		wfProfileIn( __METHOD__ );
+		$tagType = $xmlNode->getName();
+		$className = 'Wikia\\PortableInfobox\\Parser\\Nodes\\' . 'Node' . ucfirst( strtolower( $tagType ) );
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFactory.class.php]: I'd suggest using multibyte-safe functions here too

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFooter.php [hunk 8]
```diff
@@ -4,13 +4,17 @@
...
+		return $this->data;
 	}
 
-	public function isEmpty( $data ) {
-		$links = trim( $data['value'] );
+	public function isEmpty() {
+		$data = $this->getData();
+		$links = trim( $data[ 'value' ] );
+
 		return empty( $links ) && $links != '0';
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFooter.php]: I know it wasn't changed in this PR, but this line looks like duplicate logic as compared to 
https://github.com/Wikia/app/pull/7467/files#diff-b09ea7c9a303b75f3ed56e8a326e0e0cR100
if my comment there makes sense

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeImage.php [hunk 9]
```diff
@@ -8,29 +8,49 @@ class NodeImage extends Node {
...
+		}
+
+		return $this->data;
 	}
 
-	public function isEmpty( $data ) {
+	public function isEmpty() {
+		$data = $this->getData();
+
 		return !( isset( $data[ 'url' ] ) ) || empty( $data[ 'url' ] );
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeImage.php]: Do we need both sides of || here?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeSet.php [hunk 10]
```diff
@@ -1,24 +1,40 @@
...
-		return [ 'value' => $nodeFactory->getDataFromNodes( $this->xmlNode, $this ) ];
+
+		return $this->data;
+	}
+
+	public function getRenderData() {
+		return [
+			'type' => $this->getType(),
+			'data' => [
+				'value' => array_map( function ( Node $item ) {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeSet.php]: Here I think that putting each argument of array_map in its separate line will help in reading this code.

File: extensions/wikia/PortableInfobox/tests/PortableInfoboxParserTagControllerTest.php [hunk 11]
```diff
@@ -14,6 +14,11 @@ protected function setUp() {
 		$this->controller = new PortableInfoboxParserTagController();
 	}
 
+	protected function tearDown() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/PortableInfoboxParserTagControllerTest.php]: :+1:

File: extensions/wikia/PortableInfobox/tests/SimpleXmlUtilTest.php [hunk 12]
```diff
@@ -10,6 +10,12 @@ protected function setUp() {
 		$this->simpleXmlUtil = \Wikia\PortableInfobox\Helpers\SimpleXmlUtil::getInstance();
 	}
 
+	protected function tearDown() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/SimpleXmlUtilTest.php]: :+1:

File: extensions/wikia/PortableInfobox/tests/PortableInfoboxParserTagControllerTest.php (hunk 11 above)
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/PortableInfoboxParserTagControllerTest.php]: Please add a comment like in https://github.com/Wikia/app/pull/7467/files#diff-205f19f296b9012bff34f01193484beeR14

File: extensions/wikia/PortableInfobox/tests/XmlParserTest.php [hunk 13]
```diff
@@ -7,125 +7,25 @@ protected function setUp() {
...
-			}
-		));
+			->setMethods( [ 'logXmlParseError' ] )
+			->getMock();
 
-		$throwsException = false;
-
-		// getDataFromXmlString should throw an exception, but we want to proceed in order to check parameters
+		$errors = [ ];
+		// parseXmlString should throw an exception, but we want to proceed in order to check parameters
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/XmlParserTest.php]: What do you think about moving this to L24 (inside catch block) ?

File: extensions/wikia/PortableInfobox/tests/XmlParserTest.php [hunk 14]
```diff
@@ -7,125 +7,25 @@ protected function setUp() {
...
 		}
 
-		$this->assertTrue( $throwsException );
-
-		// lets compare expectedErrors with real errors
-		foreach ( $expectedErrors as $expectedError ) {
-			$this->assertTrue( in_array( $expectedError['code'], $errorCodes ) );
-			$this->assertTrue( in_array( $expectedError['msg'], $errorMsgs ) );
-		}
+		$this->assertEquals( $expectedErrors, array_map( function ( LibXMLError $error ) {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/XmlParserTest.php]: Is the order of errors deterministic?

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+	 * @param $params
+	 * @param $expected
+	 */
+	public function testGetData( $markup, $params, $expected ) {
+		$node = \Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML( $markup, $params );
+
+		$this->assertEquals( $expected, $node->getData() );
+	}
+
+	public function dataProvider() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: What about tests for an empty node and a non-empty node with empty data element?

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+	protected function setUp() {
+		$this->setupFile = dirname( __FILE__ ) . '/../../PortableInfobox.setup.php';
+		parent::setUp();
+	}
+
+	/**
+	 * @covers       NodeComparison::getData
+	 * @covers       Node::getDataForChildren
+	 * @covers       Node::getChildNodes
+	 * @dataProvider dataProvider
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: Nitpick: getDataDataProvider? :)

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	 * @param string $markup XML to test
+	 * @param array $expected
+	 */
+	public function testSource( $markup, $expected ) {
+		$node = \Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML( $markup, [ ] );
+
+		$this->assertEquals( $expected, $node->getSource() );
+	}
+
+	public function sourceDataProvider() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: A test set with duplicate source value would be nice too

File: extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+	 * @param $params
+	 * @param $expected
+	 */
+	public function testSource( $markup, $params, $expected ) {
+		$node = \Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML( $markup, $params );
+
+		$this->assertEquals( $expected, $node->getSource() );
+	}
+
+	public function sourceDataProvider() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeDataTest.php]: Also, please add duplicate source test

File: extensions/wikia/PortableInfobox/tests/nodes/NodeImageTest.php [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+	 * @param $markup
+	 * @param $expected
+	 */
+	public function testSource( $markup, $expected ) {
+		$node = \Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML( $markup, [ ] );
+
+		$this->assertEquals( $expected, $node->getSource() );
+	}
+
+	public function sourceProvider() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeImageTest.php]: And here too (duplicate source param test case)

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeComparison.php [hunk 20]
```diff
@@ -1,24 +1,37 @@
...
-		return [ 'value' => $nodeFactory->getDataFromNodes( $this->xmlNode, $this ) ];
+
+		return $this->data;
+	}
+
+	public function getRenderData() {
+		return [
+			'type' => $this->getType(),
+			'data' => [ 'value' => $this->getRenderDataForChildren() ],
+			'isEmpty' => $this->isEmpty(),
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeComparison.php]: Per offline discussion, we can get rid of this and the below line

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+	 * @param $params
+	 * @param $expected
+	 */
+	public function testRenderData( $markup, $params, $expected ) {
+		$node = \Wikia\PortableInfobox\Parser\Nodes\NodeFactory::newFromXML( $markup, $params );
+
+		$this->assertEquals( $expected, $node->getRenderData() );
+	}
+
+	public function renderDataProvider() {
```
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: Please add tests covering getRenderData in context of an empty node for all nodes

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php (hunk 3 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: We use builder pattern here, for fluently passing external parser to children nodes. Type hinting was removed to prevent catchable fatal error appearing.

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFooter.php (hunk 8 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/NodeFooter.php]: lets leave it for now, we should go through isEmpty logic one more time later in time

File: extensions/wikia/PortableInfobox/tests/XmlParserTest.php (hunk 14 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/tests/XmlParserTest.php]: yes

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php (hunk 16 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: please no

Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: Is there something less generic than dataProvider in that case?

File: extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php (hunk 3 above)
Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/services/Parser/Nodes/Node.php]: Let's add this info to the comment

File: extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php (hunk 21 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/tests/nodes/NodeComparisonTest.php]: This should be covered in parent node

File: extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php (hunk 1 above)
Comment (idradm) [inline extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php]: ib stand for infobox, this is an params prefix, taking into account we don't use params, maybe we shouldn't worry about that

Comment (SebastianMarzjan) [inline extensions/wikia/PortableInfobox/controllers/ApiQueryPortableInfobox.class.php]: Seems fair.

**Comment by SebastianMarzjan** [comment]
lgtm
